test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go

```diff
@@ -0,0 +1,508 @@
+		},
+		predicates.ClusterUnpausedAndInfrastructureReady(r.Log),
+	)
+}
+
+func (r *DockerMachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, machinePool *clusterv1exp.MachinePool, dockerMachinePool *infrav1exp.DockerMachinePool, log logr.Logger) (ctrl.Result, error) {
+	pool, err := newNodePool(r, cluster, machinePool, dockerMachinePool, log)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to build new node pool")
+	}
+
+	if err := pool.Delete(ctx); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to delete all machines in the node pool")
+	}
+
+	controllerutil.RemoveFinalizer(dockerMachinePool, infrav1exp.MachinePoolFinalizer)
+	return ctrl.Result{}, nil
+}
+
+func (r *DockerMachinePoolReconciler) reconcileNormal(ctx context.Context, cluster *clusterv1.Cluster, machinePool *clusterv1exp.MachinePool, dockerMachinePool *infrav1exp.DockerMachinePool, log logr.Logger) (ctrl.Result, error) {
+	// Make sure bootstrap data is available and populated.
+	if machinePool.Spec.Template.Spec.Bootstrap.DataSecretName == nil {
+		log.Info("Waiting for the Bootstrap provider controller to set bootstrap data")
+		return ctrl.Result{}, nil
+	}
+
+	if machinePool.Spec.Replicas == nil {
+		machinePool.Spec.Replicas = pointer.Int32Ptr(1)
+	}
+
+	pool, err := newNodePool(r, cluster, machinePool, dockerMachinePool, log)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to build new node pool")
+	}
+
+	// if we don't have enough nodes matching spec, build them
+	if err := pool.ReconcileMachines(ctx); err != nil {
+		if errors.Is(err, &TransientError{}) {
+			log.V(4).Info("requeue in 5 seconds due docker machine reconcile transient error")
+			return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+		}
+
+		return ctrl.Result{}, errors.Wrap(err, "failed to reconcile machines")
+	}
+
+	if err := pool.DeleteExtraMachines(ctx); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to delete overprovisioned or out of spec machines")
+	}
+
+	dockerMachinePool.Spec.ProviderIDList = []string{}
+	for _, instance := range dockerMachinePool.Status.Instances {
+		if instance.ProviderID != nil && instance.Ready {
+			dockerMachinePool.Spec.ProviderIDList = append(dockerMachinePool.Spec.ProviderIDList, *instance.ProviderID)
+		}
+	}
+
+	dockerMachinePool.Status.Replicas = int32(len(dockerMachinePool.Spec.ProviderIDList))
+
+	if dockerMachinePool.Spec.ProviderID == "" {
+		dockerMachinePool.Spec.ProviderID = getDockerMachinePoolProviderID(cluster.Name, dockerMachinePool.Name)
+	}
+
+	dockerMachinePool.Status.Ready = len(dockerMachinePool.Spec.ProviderIDList) == int(*machinePool.Spec.Replicas)
+	return ctrl.Result{}, nil
+}
+
+type nodePool struct {
+	client            client.Client
+	cluster           *clusterv1.Cluster
+	machinePool       *clusterv1exp.MachinePool
+	dockerMachinePool *infrav1exp.DockerMachinePool
+	machines          []*docker.Machine
```

Comment:
So this is using `InfraMachines` underneath, which makes sense since docker infra doesn't have a scale set concept, but maybe we should document either in code comments or in the docker exp documentation that this is a workaround for docker infra because it doesn't have a native infra resource for pool? I'm just worried about new infra providers using docker as the reference infra provider and getting misled into thinking this is how MachinePools should be implemented on the infra side.

I haven't gone through the whole PR yet so maybe this is already done.

Reply:
I've broken out the node pool and added docs for the type / funcs.